### PR TITLE
Date axis support for Vertical and Vertical Stacked Bar charts

### PR DESCRIPTION
**Pull request by ankityadav4**
## New Behavior
image

image

## Related Issue(s)

- Fixes #

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/419d7609ad9ce151f8f544c891d55e3a395c13d1'>419d7609ad9ce151f8f544c891d55e3a395c13d1</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv8 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v8"/>

File: packages/react-charting/src/types/IDataPoint.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ export interface IBasestate {
   refSelected?: any;
   YValueHover?: { legend?: string; y?: number; color?: string }[];
   hoverYValue?: string | number | null;
-  hoverXValue?: string | number | null;
+  hoverXValue?: string | number | Date | null;
```
Comment (AtishayMsft) [inline packages/react-charting/src/types/IDataPoint.ts:17]: Since this is a common variable across all charts, are there charts where the Date type won't work.

File: packages/react-charting/src/utilities/utilities.ts [hunk 2]
```diff
@@ -1004,6 +1005,85 @@ export function domainRangeOfVSBCNumeric(
...
+  const rStartValue = margins.left! + barWidth / 2;
+  const rEndValue = width - margins.right! - barWidth / 2;
+
+  return isRTL
+    ? { dStartValue: lDate, dEndValue: sDate, rStartValue, rEndValue }
+    : { dStartValue: sDate, dEndValue: lDate, rStartValue, rEndValue };
+}
+
+/**
+ * Calculate domain and range values to the Vertical stacked bar chart - For Date axis
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/utilities.ts:1049]: how to avoid the code duplication here?

File: packages/react-charting/src/utilities/utilities.ts [hunk 3]
```diff
@@ -1082,6 +1162,26 @@ export function getDomainNRangeValues(
       case ChartTypes.LineChart:
         domainNRangeValue = domainRangeOfDateForAreaChart(points, margins, width, isRTL, tickValues! as Date[]);
         break;
+      case ChartTypes.VerticalBarChart:
+        domainNRangeValue = domainRangeOfDateForVerticalBarChart(
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/utilities.ts:1166]: Can't we use the domainRangeOfDateForAreaChart function for vertical bar and vertical stacked use case also.

File: packages/react-charting/src/utilities/utilities.ts [hunk 4]
```diff
@@ -1350,10 +1450,11 @@ export const convertToLocaleString = (data: LocaleStringDataProps, culture?: str
   culture = culture || undefined;
   if (typeof data === 'number') {
     return data.toLocaleString(culture);
-  }
-  if (typeof data === 'string' && !window.isNaN(Number(data))) {
+  } else if (typeof data === 'string' && !window.isNaN(Number(data))) {
     const num = Number(data);
     return num.toLocaleString(culture);
+  } else if (data instanceof Date) {
+    return data.toLocaleDateString(culture);
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/utilities.ts:1457]: why is this change needed. It might break existing charts.

File: packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts [hunk 5]
```diff
@@ -486,7 +486,7 @@ export interface IModifiedCartesianChartProps extends ICartesianChartProps {
     isCalloutVisible: boolean;
     id: string;
     YValueHover?: IYValueHover[];
-    hoverXValue?: string | number | null;
+    hoverXValue?: string | number | Date | null;
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts:489]: Is there no chart for which hoverXValue = Date does not work now

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx [hunk 6]
```diff
@@ -511,26 +517,44 @@ export class VerticalBarChartBase extends React.Component<IVerticalBarChartProps
...
+        return point.x as Date;
+      })!;
+      const xBarScale = d3ScaleTime()
+        .domain([sDate, lDate])
+        .range([
+          this.margins.left! + this._domainMargin + this._barWidth / 2,
+          containerWidth - this.margins.right! - this._barWidth / 2 - this._domainMargin,
+        ]);
 
       const yBarScale = d3ScaleLinear()
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx:535]: >yBarScale [](http://example.com/codeflow?start=12&length=9)

yBarScale code is repeating thrice. We should have it only once.

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx [hunk 7]
```diff
@@ -542,7 +566,7 @@ export class VerticalBarChartBase extends React.Component<IVerticalBarChartProps
       const barHeight: number = Math.max(yBarScale(point.y), 0);
       let adjustedBarHeight = 0;
       if (barHeight <= 0) {
-        return <React.Fragment key={point.x}> </React.Fragment>;
+        return <React.Fragment key={point.x instanceof Date ? point.x.getTime() : point.x}> </React.Fragment>;
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx:569]: >instanceof [](http://example.com/codeflow?start=44&length=10)

This will always be numeric as it is part of the numeric function

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx [hunk 8]
```diff
@@ -815,7 +918,8 @@ export class VerticalBarChartBase extends React.Component<IVerticalBarChartProps
   }
 
   private _getDomainMargins = (containerWidth: number): IMargins => {
-    if (this._xAxisType !== XAxisTypes.NumericAxis) {
+    if (this._xAxisType !== XAxisTypes.NumericAxis && this._xAxisType !== XAxisTypes.DateAxis) {
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx:921]: >this [](http://example.com/codeflow?start=8&length=4)

Add a positive check for string axis.

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx [hunk 9]
```diff
@@ -601,13 +625,13 @@ export class VerticalBarChartBase extends React.Component<IVerticalBarChartProps
   }
 
   private _createStringBars(containerHeight: number, containerWidth: number, xElement: SVGElement): JSX.Element[] {
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx:627]: >_createStringBars [](http://example.com/codeflow?start=10&length=17)

Can we reduce code duplication in these 3 functions creating bar for numeric/string/date.

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx [hunk 10]
```diff
@@ -79,7 +84,7 @@ export class VerticalStackedBarChartBase extends React.Component<
   IVerticalStackedBarChartState
 > {
   private _points: IVerticalStackedChartProps[];
-  private _dataset: IDataPoint[];
+  private _dataset: IVerticalStackedBarDataPoint[];
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:87]: >IVerticalStackedBarDataPoint [](http://example.com/codeflow?start=20&length=28)

Will this break existing users?

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx [hunk 11]
```diff
@@ -314,20 +320,30 @@ export class VerticalStackedBarChartBase extends React.Component<
...
           : // eslint-disable-next-line @typescript-eslint/no-explicit-any
             (xBarScale as any)(lineObject[item][i - 1].xItem.xAxisPoint as string);
         const useSecondaryYScale =
           lineObject[item][i - 1].useSecondaryYScale && lineObject[item][i].useSecondaryYScale && secondaryYScale;
         const y1 = useSecondaryYScale ? secondaryYScale!(lineObject[item][i - 1].y) : yScale(lineObject[item][i - 1].y);
         const x2 = isNumeric
           ? xScale(lineObject[item][i].xItem.xAxisPoint as number)
+          : isDate
+          ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            (xBarScale as any)(lineObject[item][i].xItem.xAxisPoint as Date)
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:346]: >any [](http://example.com/codeflow?start=26&length=3)

why using any here

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx [hunk 12]
```diff
@@ -748,10 +767,17 @@ export class VerticalStackedBarChartBase extends React.Component<
...
         this._xAxisType === XAxisTypes.NumericAxis
           ? (singleChartData.xAxisPoint as number)
+          : this._xAxisType === XAxisTypes.DateAxis
+          ? (singleChartData.xAxisPoint as Date)
           : (singleChartData.xAxisPoint as string),
       );
       const xScaleBandwidthTranslate =
-        this._xAxisType === XAxisTypes.NumericAxis ? 0 : (xBarScale.bandwidth() - this._barWidth) / 2;
+        this._xAxisType === XAxisTypes.NumericAxis
+          ? 0
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:776]: >0 [](http://example.com/codeflow?start=12&length=1)

shouldn't this behave same as numeric scale

File: packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.DateLabel.Example.tsx [hunk 13]
```diff
@@ -0,0 +1,70 @@
+import * as React from 'react';
+import { VerticalBarChart, IVerticalBarChartProps, IVerticalBarChartDataPoint } from '@fluentui/react-charting';
+
+interface IVerticalBarState {}
+
+export class VerticalBarChartDateLabelExample extends React.Component<{}, IVerticalBarState> {
```
Comment (AtishayMsft) [inline packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.DateLabel.Example.tsx:6]: >VerticalBarChartDateLabelExample [](http://example.com/codeflow?start=13&length=32)

VerticalBarChartDateAxisExample

File: packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChartPage.tsx [hunk 14]
```diff
@@ -48,6 +51,9 @@ export class VerticalBarChartPage extends React.Component<IComponentDemoPageProp
             <ExampleCard title="VerticalBarChart dynamic" code={VerticalBarChartTooltipExampleCode}>
               <VerticalBarChartTooltipExample />
             </ExampleCard>
+            <ExampleCard title="VerticalBarChart Date Axis" code={VerticalBarChartDateLabelExampleCode}>
```
Comment (AtishayMsft) [inline packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChartPage.tsx:54]: You need to add an entry in VerticalBarChart.doc.tsx to show in the docsite.
Also update the supported axis table in docs/TechnicalDetails.md

File: packages/react-charting/src/utilities/utilities.ts (hunk 4 above)
Comment (AtishayMsft) [reply packages/react-charting/src/utilities/utilities.ts:1457]: The hover label is coming very long.

File: change/@fluentui-react-charting-90e2ea85-e637-474e-8689-217a35791237.json [hunk 15]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
```
[the remark this answers is not in the document]
Comment (ankityadav4) [reply change/@fluentui-react-charting-90e2ea85-e637-474e-8689-217a35791237.json:2]: Done

File: packages/react-charting/src/types/IDataPoint.ts (hunk 1 above)
Comment (ankityadav4) [reply packages/react-charting/src/types/IDataPoint.ts]: This is not needed anymore.

File: packages/react-charting/src/utilities/utilities.ts (hunk 2 above)
Comment (ankityadav4) [reply packages/react-charting/src/utilities/utilities.ts]: Done

File: packages/react-charting/src/utilities/utilities.ts (hunk 3 above)
Comment (ankityadav4) [reply packages/react-charting/src/utilities/utilities.ts]: Done

File: packages/react-charting/src/utilities/utilities.ts (hunk 4 above)
Comment (ankityadav4) [reply packages/react-charting/src/utilities/utilities.ts:1449]: Hover length is now proper.

File: packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts (hunk 5 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts]: Reverted this change

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx (hunk 6 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx]: Done

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx (hunk 7 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx]: This is fixed

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 10 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:87]: It is type of IDataPoint with x redeclared as Date type also. It will not break existing users.This is done because IDataPoint is used in other charts.

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 12 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:776]: I tried same using similar logic as numeric scale but bars are not coming properly. Time Range is also +/-barwidth/2 , so we have to adjust that while making bars.

File: packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.DateLabel.Example.tsx (hunk 13 above)
Comment (ankityadav4) [reply packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.DateLabel.Example.tsx]: Done

File: packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChartPage.tsx (hunk 14 above)
Comment (ankityadav4) [reply packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChartPage.tsx]: Done

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx (hunk 9 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx:624]: This i will take up in separate PR.

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 11 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:346]: Because xBarScale can be of type d3ScaleLinear | d3ScaleTime | d3ScaleBand , and the condition check is on point.x type and not on the xBarScale type

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 12 above)
Comment (AtishayMsft) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:776]: could you mention this as comment in the code

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx [hunk 16]
```diff
@@ -201,8 +205,8 @@ export class VerticalBarChartBase extends React.Component<IVerticalBarChartProps
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     yScaleSecondary?: any,
   ): React.ReactNode => {
-    const isNumericAxis = this._xAxisType === XAxisTypes.NumericAxis;
-    const { xBarScale } = this._getScales(containerHeight, containerWidth, isNumericAxis);
+    const isStringAxis = this._xAxisType !== XAxisTypes.DateAxis && this._xAxisType !== XAxisTypes.NumericAxis;
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx:208]: >     [](http://example.com/codeflow?start=0&length=4)

just check if axistype === string axis

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx [hunk 17]
```diff
@@ -815,7 +915,8 @@ export class VerticalBarChartBase extends React.Component<IVerticalBarChartProps
   }
 
   private _getDomainMargins = (containerWidth: number): IMargins => {
-    if (this._xAxisType !== XAxisTypes.NumericAxis) {
+    if (this._xAxisType === XAxisTypes.StringAxis) {
+      /*) {
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx:919]: remove this extra line

File: packages/react-charting/src/utilities/utilities.ts [hunk 18]
```diff
@@ -1087,7 +1143,36 @@ export function getDomainNRangeValues(
     switch (chartType) {
       case ChartTypes.AreaChart:
       case ChartTypes.LineChart:
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/utilities.ts:1145]: >ChartTypes [](http://example.com/codeflow?start=11&length=10)

this can be a single group of cases for area, line, vbc, stacked vbc

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 10 above)
Comment (AtishayMsft) [inline packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:87]: >private [](http://example.com/codeflow?start=2&length=7)

can you try with an example using IDataPoint in the input data.
Does it throw a warning or lint error or something

File: packages/react-charting/src/utilities/utilities.ts (hunk 2 above)
Comment (AtishayMsft) [reply packages/react-charting/src/utilities/utilities.ts]: You can resolve the comments that have been addressed.

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 12 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:778]: Done

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx (hunk 16 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx]: Done

File: packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx (hunk 17 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalBarChart/VerticalBarChart.base.tsx]: Done

File: packages/react-charting/src/utilities/utilities.ts (hunk 18 above)
Comment (ankityadav4) [reply packages/react-charting/src/utilities/utilities.ts:1145]: Done

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 10 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:87]: This is not exposed in examples.

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx [hunk 19]
```diff
@@ -964,10 +985,8 @@ export class VerticalStackedBarChartBase extends React.Component<
     containerWidth: number,
     xElement: SVGElement | null,
   ) => {
-    return (this._bars =
-      this._xAxisType === XAxisTypes.NumericAxis
-        ? this._createNumericBars(containerHeight, containerWidth, xElement!)
-        : this._createStringBars(containerHeight, containerWidth, xElement!));
+    const { xBarScale, yBarScale } = this._getScales(containerHeight, containerWidth);
+    return (this._bars = this._createBar(xBarScale, yBarScale, containerHeight, xElement!));
```
[the remark this answers is not in the document]
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:991]: I am not able to open this link.

File: packages/react-charting/src/utilities/utilities.ts [hunk 20]
```diff
@@ -1087,7 +1143,17 @@ export function getDomainNRangeValues(
...
-        domainNRangeValue = domainRangeOfDateForAreaChart(points, margins, width, isRTL, tickValues! as Date[]);
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/utilities.ts:1090]: this function can be removed

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 10 above)
Comment (AtishayMsft) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:87]: Then this interface should not be exposed publicly in the api.md file. It only confuses the user. Can you log a task for it.

**Review by AtishayMsft — APPROVED** [review]

File: packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 10 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:87]: Created the task for it.

File: packages/react-charting/src/utilities/utilities.ts (hunk 20 above)
Comment (ankityadav4) [reply packages/react-charting/src/utilities/utilities.ts:1090]: Done.
